relentless/_collections.py

```diff
@@ -266,3 +268,254 @@ def __str__(self):
     def pairs(self):
         """tuple: All unique pairs in the matrix."""
         return tuple(self._data.keys())
+
+class KeyedArray(FixedKeyDict):
+    """Numerical array with fixed keys.
+
+    Can be used to perform arithmetic operations between two arrays (element-wise)
+    or between an array and a scalar, as well as vector algebraic operations
+    (norm, dot product).
+
+    Parameters
+    ----------
+    keys : array_like
+        List of keys to be fixed.
+    default : scalar
+        Initial value to fill in the dictionary, defaults to `None`.
+
+    Examples
+    --------
+    Create a keyed array::
+
+        k1 = KeyedArray(keys=('A','B'))
+        k2 = KeyedArray(keys=('A','B'))
+
+    Set values through update::
+
+        k1.update({'A':2.0, 'B':3.0})
+        k2.update({'A':3.0, 'B':4.0})
+
+    Perform array-array arithmetic operations::
+
+        >>> print(k1 + k2)
+        {'A':5.0, 'B':7.0}
+        >>> print(k1 - k2)
+        {'A':-1.0, 'B':-1.0}
+        >>> print(k1*k2)
+        {'A':6.0, 'B':12.0}
+        >>> print(k1/k2)
+        {'A':0.6666666666666666, 'B':0.75}
+
+    Perform array-scalar arithmetic operations::
+
+        >>> print(k1 + 3)
+        {'A':5.0, 'B':6.0}
+        >>> print(3 - k1)
+        {'A':1.0, 'B':0.0}
+        >>> print(3*k1)
+        {'A':6.0, 'B':9.0}
+        >>> print(k1/10)
+        {'A':0.2, 'B':0.3}
+
+    Compute vector dot product::
+
+        >>> print(k1.dot(k2))
+        18.0
+
+    Compute vector norm::
+
+        >>> print(k2.norm())
+        5.0
+
+    """
+    def __init__(self, keys, default=None):
+        super().__init__(keys, default)
+
+    def _assert_same_keys(self, val):
+        if (self.keys != val.keys):
+            raise KeyError('Both KeyedArrays must have identical keys to perform mathematical operations.')
+
+    def __add__(self, val):
+        """Element-wise addition of two arrays, or of an array and a scalar."""
+        k = KeyedArray(keys=self.keys)
+        if isinstance(val, KeyedArray):
+            self._assert_same_keys(val)
+            k.update({x: self[x] + val[x] for x in self})
+        elif np.isscalar(val):
+            k.update({x: self[x] + val for x in self})
+        else:
+            raise TypeError('A KeyedArray can only add a scalar or a KeyedArray.')
+        return k
+
+    def __radd__(self, val):
```

Comment:
Just to check, you only need to implement `__radd__` for the case where the left `val` is a scalar because `__add__` will be invoked if the `val` is a `KeyedArray`?

Reply:
Yes!